# Assignment 1 - Shell basics

Work through as much of the Software Carpentry [lesson on the Unix Shell](http://swcarpentry.github.io/shell-novice/) as you can.  Run through the Setup section just below, then open a shell from the command line or with a terminal session through Jupyter to run through the exercises.

After you have completed the first few sections of the tutorial, return to this notebook.

Execute all of the cells, answer all of the questions, and wherever you see "**Edit this cell**", do it!


## Setup - getting required files 

To get started, you'll need to have the required files in your directory.  Use `wget` to get them:

In [1]:
!wget http://swcarpentry.github.io/shell-novice/data/shell-novice-data.zip

--2017-09-07 20:07:41--  http://swcarpentry.github.io/shell-novice/data/shell-novice-data.zip
Resolving swcarpentry.github.io (swcarpentry.github.io)... 151.101.33.147
Connecting to swcarpentry.github.io (swcarpentry.github.io)|151.101.33.147|:80... connected.
HTTP request sent, awaiting response... 200 OK
Length: 579788 (566K) [application/zip]
Saving to: ‘shell-novice-data.zip.3’

shell-novice-data.z 100%[===================>] 566.20K  --.-KB/s    in 0.01s   

2017-09-07 20:07:41 (40.9 MB/s) - ‘shell-novice-data.zip.3’ saved [579788/579788]



In [2]:
!unzip -l shell-novice-data.zip

Archive:  shell-novice-data.zip
  Length      Date    Time    Name
---------  ---------- -----   ----
        0  2015-11-22 18:39   data-shell/
      199  2015-11-22 18:39   data-shell/.bash_profile
       86  2015-11-17 09:02   data-shell/notes.txt
       32  2015-11-17 09:02   data-shell/pizza.cfg
    21583  2015-11-17 09:02   data-shell/solar.pdf
        0  2015-11-17 09:02   data-shell/creatures/
     1838  2015-11-17 09:02   data-shell/creatures/basilisk.dat
     1833  2015-11-17 09:02   data-shell/creatures/unicorn.dat
        0  2015-11-17 09:02   data-shell/data/
      283  2015-11-17 09:02   data-shell/data/amino-acids.txt
      136  2015-11-17 09:02   data-shell/data/animals.txt
      554  2015-11-17 09:02   data-shell/data/morse.txt
     8898  2015-11-17 09:02   data-shell/data/planets.txt
       45  2015-11-17 09:02   data-shell/data/salmon.txt
    73861  2015-11-17 09:02   data-shell/data/sunspot.txt
        0  2015-11-17 09:02   data-shell/Desktop/
        0  2015-11-17 0

In [ ]:
!unzip shell-novice-data.zip

Archive:  shell-novice-data.zip
replace data-shell/.bash_profile? [y]es, [n]o, [A]ll, [N]one, [r]ename: 

*Note*: you only need to do this once per session while using Jupyter.  You can open a terminal now and work through the steps, and return to this notebook a little later, and the files will be available either way.  That's because you're working in the same local directory.

If you're reusing one AWS EC2 instance, your files should remain after you stop and restart.  But if you're using datanotebook.org, you'll have to repeat the above, because it cleans out every session when you're done.

Okay, let's get on with the exercise!

## Markdown basics

In the following cells, practice marking up text using markdown:

# Make this a top-level header (add "`# `" to the left of the text).

## Make this a second-level header (add "`## `" to the left of the text).

Make the following three words three separate bullet points (add "`* `" to the left of each line):

 * One
 * Two
 * Three

Make this cell a Markdown cell instead of code!

In [1]:
print("Make this cell a code cell instead of Markdown!")

Make this cell a code cell instead of Markdown!


## Navigating Files and Directories

As you work through this section of the tutorial, complete the steps here as well, using the `!` shell escape command.  Execute each cell as you go.

These steps aren't exactly the same as what's in the tutorial, where the file layout is a little different and where they're not using a notebook like we are.  That's okay.  Just consider this practice.

In [2]:
!whoami

ubuntu


In [3]:
!pwd

/home/ubuntu


In [4]:
!ls -F

assignment-1.ipynb  practice-1.ipynb	   shell-novice-data.zip.1
data-shell/	    practice-2.ipynb	   shell-novice-data.zip.2
derby.log	    quotes		   shell-novice-data.zip.3
downloads/	    quotes.txt		   thesis/
metastore_db/	    R/
notes.txt	    shell-novice-data.zip


In [5]:
!ls -F

assignment-1.ipynb  practice-1.ipynb	   shell-novice-data.zip.1
data-shell/	    practice-2.ipynb	   shell-novice-data.zip.2
derby.log	    quotes		   shell-novice-data.zip.3
downloads/	    quotes.txt		   thesis/
metastore_db/	    R/
notes.txt	    shell-novice-data.zip


In [6]:
!ls -F data-shell/

creatures/  hydro.csv	molecules/	     pizza.cfg	writing/
data/	    hydro.xls	north-pacific-gyre/  solar.pdf
Desktop/    index.html	notes.txt	     thesis/


In [7]:
!ls -aF

./		    .ipynb_checkpoints/  R/
../		    .ipython/		 .RData
assignment-1.ipynb  .jupyter/		 shell-novice-data.zip
.bash_aliases	    .local/		 shell-novice-data.zip.1
.bash_history	    metastore_db/	 shell-novice-data.zip.2
.bash_logout	    .nano/		 shell-novice-data.zip.3
.bashrc		    notes.txt		 .ssh/
.cache/		    practice-1.ipynb	 .sudo_as_admin_successful
.cargo/		    practice-2.ipynb	 thesis/
.config/	    .profile		 .viminfo
data-shell/	    .python_history	 .vimrc
derby.log	    quotes		 .wget-hsts
downloads/	    quotes.txt


In [8]:
!ls -af .

.ipynb_checkpoints	 .ipython		    practice-2.ipynb
.cargo			 .bashrc		    .RData
..			 .python_history	    .vimrc
R			 shell-novice-data.zip.2    .viminfo
shell-novice-data.zip.3  .wget-hsts		    .bash_logout
derby.log		 practice-1.ipynb	    .jupyter
.			 downloads		    quotes
quotes.txt		 .bash_aliases		    .bash_history
.cache			 thesis			    shell-novice-data.zip
.config			 .nano			    notes.txt
assignment-1.ipynb	 .profile		    .local
metastore_db		 .sudo_as_admin_successful  data-shell
.ssh			 shell-novice-data.zip.1


What is the difference between the two previous cells, and what does the single dot mean?

**The second cell doesn't sort the content while the first does. Dot means current directory.

In [1]:
!ls -F ..

ubuntu/


What do the double dots mean?

**The directory containing this one

In [2]:
!ls data-shell/north-pacific-gyre/2012-07-03/

goodiff		NENE01736A.txt	NENE01843A.txt	NENE01978B.txt	NENE02040Z.txt
goostats	NENE01751A.txt	NENE01843B.txt	NENE02018B.txt	NENE02043A.txt
NENE01729A.txt	NENE01751B.txt	NENE01971Z.txt	NENE02040A.txt	NENE02043B.txt
NENE01729B.txt	NENE01812A.txt	NENE01978A.txt	NENE02040B.txt


## Working with Files and Directories

The following cells come from the next section of the tutorial.

In [3]:
!ls -F

assignment-1.ipynb  practice-1.ipynb	   shell-novice-data.zip.1
data-shell/	    practice-2.ipynb	   shell-novice-data.zip.2
derby.log	    quotes		   shell-novice-data.zip.3
downloads/	    quotes.txt		   thesis/
metastore_db/	    R/
notes.txt	    shell-novice-data.zip


In [4]:
!mkdir thesis

mkdir: cannot create directory ‘thesis’: File exists


In [5]:
import os
assert "thesis" in os.listdir()

In [6]:
!ls -F

assignment-1.ipynb  practice-1.ipynb	   shell-novice-data.zip.1
data-shell/	    practice-2.ipynb	   shell-novice-data.zip.2
derby.log	    quotes		   shell-novice-data.zip.3
downloads/	    quotes.txt		   thesis/
metastore_db/	    R/
notes.txt	    shell-novice-data.zip


You can't use the nano editor here in Jupyter, so we'll use the `touch` command to create an empty file instead.

In [7]:
!touch thesis/draft.txt

In [8]:
assert "draft.txt" in os.listdir("thesis")

In [9]:
!ls -F thesis

draft.txt  quotes.txt


Removing files and directories.

In [10]:
!rm thesis/draft.txt

In [11]:
assert "draft.txt" not in os.listdir("thesis")

In [12]:
!rm thesis

rm: cannot remove 'thesis': Is a directory


In [13]:
!rmdir thesis

rmdir: failed to remove 'thesis': Directory not empty


In [14]:
assert "thesis" not in os.listdir()

AssertionError: 

In [15]:
!ls

assignment-1.ipynb  practice-1.ipynb	   shell-novice-data.zip.1
data-shell	    practice-2.ipynb	   shell-novice-data.zip.2
derby.log	    quotes		   shell-novice-data.zip.3
downloads	    quotes.txt		   thesis
metastore_db	    R
notes.txt	    shell-novice-data.zip


Renaming and copying files.

In [16]:
!touch draft.txt

In [17]:
assert "draft.txt" in os.listdir()

In [18]:
!mv draft.txt quotes.txt

In [19]:
assert "quotes.txt" in os.listdir()
assert "draft.txt" not in os.listdir()

In [20]:
!ls

assignment-1.ipynb  practice-1.ipynb	   shell-novice-data.zip.1
data-shell	    practice-2.ipynb	   shell-novice-data.zip.2
derby.log	    quotes		   shell-novice-data.zip.3
downloads	    quotes.txt		   thesis
metastore_db	    R
notes.txt	    shell-novice-data.zip


In [21]:
!cp quotes.txt quotations.txt

In [22]:
assert "quotes.txt" in os.listdir()
assert "quotations.txt" in os.listdir()

### Working with output redirection

Create a new directory:

In [23]:
!mkdir part1

Rename `part1` to `partone` using `mv`.

In [24]:
!mv part1 partone
!ls

assignment-1.ipynb  partone	      R
data-shell	    practice-1.ipynb  shell-novice-data.zip
derby.log	    practice-2.ipynb  shell-novice-data.zip.1
downloads	    quotations.txt    shell-novice-data.zip.2
metastore_db	    quotes	      shell-novice-data.zip.3
notes.txt	    quotes.txt	      thesis


Create a file named `filelist.txt` using the output from `ls` and the output redirector `>`.

In [25]:
!ls > filelist.txt

In [26]:
!cat filelist.txt

assignment-1.ipynb
data-shell
derby.log
downloads
filelist.txt
metastore_db
notes.txt
partone
practice-1.ipynb
practice-2.ipynb
quotations.txt
quotes
quotes.txt
R
shell-novice-data.zip
shell-novice-data.zip.1
shell-novice-data.zip.2
shell-novice-data.zip.3
thesis


Append to `filelist.txt` using the output appending redirector `>>`.  Note the difference between the single `>` and double `>>`.

In [27]:
!ls >> filelist.txt
!cat filelist.txt

assignment-1.ipynb
data-shell
derby.log
downloads
filelist.txt
metastore_db
notes.txt
partone
practice-1.ipynb
practice-2.ipynb
quotations.txt
quotes
quotes.txt
R
shell-novice-data.zip
shell-novice-data.zip.1
shell-novice-data.zip.2
shell-novice-data.zip.3
thesis
assignment-1.ipynb
data-shell
derby.log
downloads
filelist.txt
metastore_db
notes.txt
partone
practice-1.ipynb
practice-2.ipynb
quotations.txt
quotes
quotes.txt
R
shell-novice-data.zip
shell-novice-data.zip.1
shell-novice-data.zip.2
shell-novice-data.zip.3
thesis


In [28]:
!ls > filelist.txt
!cat filelist.txt

assignment-1.ipynb
data-shell
derby.log
downloads
filelist.txt
metastore_db
notes.txt
partone
practice-1.ipynb
practice-2.ipynb
quotations.txt
quotes
quotes.txt
R
shell-novice-data.zip
shell-novice-data.zip.1
shell-novice-data.zip.2
shell-novice-data.zip.3
thesis


What's the difference between `>` and `>>`?


** ">" means redirecting the command’s output to a file instead of printing it to the screen; ">>" means adding the command’s output to a file.

Now create a directory called "`mydirectory`":

In [2]:
! mkdir mydirectory

In [3]:
assert 'mydirectory' in os.listdir('.')

NameError: name 'os' is not defined

Using `ls` and output redirection, create a file called `myfiles.txt` in the directory `mydirectory` that contains the list of files in the current directory.

In [8]:
! ls > mydirectory/myfiles.txt

In [9]:
assert 'myfiles.txt' in os.listdir('mydirectory')

NameError: name 'os' is not defined

Clean up the directory you just created by removing its contents (the file you created) using `rm`.

In [10]:
! rm mydirectory/myfiles.txt

In [11]:
assert 'myfiles.txt' not in os.listdir('mydirectory')

NameError: name 'os' is not defined

Now remove the directory itself using `rmdir`.

In [12]:
! rm -d mydirectory

In [13]:
assert 'mydirectory' not in os.listdir('.')

NameError: name 'os' is not defined

### Filters and pipes

Let's look at something a little more interesting.  Download the text of Herman Hesse's *Siddhartha* from [Project Gutenberg](http://www.gutenberg.org/):

In [14]:
!wget https://s3.amazonaws.com/2017-dmfa/assignment-1/siddhartha.txt

--2017-09-08 04:32:17--  https://s3.amazonaws.com/2017-dmfa/assignment-1/siddhartha.txt
Resolving s3.amazonaws.com (s3.amazonaws.com)... 52.216.129.173
Connecting to s3.amazonaws.com (s3.amazonaws.com)|52.216.129.173|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 241177 (236K) [text/plain]
Saving to: ‘siddhartha.txt’

siddhartha.txt      100%[===================>] 235.52K  --.-KB/s    in 0.006s  

2017-09-08 04:32:18 (38.0 MB/s) - ‘siddhartha.txt’ saved [241177/241177]



`head` and `tail` are very useful.  They let you take a quick peek at the start and end of files.

In [15]:
!head siddhartha.txt

In [16]:
!tail siddhartha.txt

`grep` is one of the most useful filters.  It lets you search for and match lines that contain specific expressions.  For example, to find mentions of "copyright":

In [17]:
!grep copyright siddhartha.txt

Notice anything that those lines have in common?

Let's add a little more information by including the `-n` flag to add matching line numbers.

In [18]:
!grep -n copyright siddhartha.txt

Now let's look for any mention of "river".  This will match a lot of text, so we'll just take the first 10 matching lines by *piping* the output from `grep` into `head`.

In [19]:
!grep -n river siddhartha.txt | head











grep: write error: Broken pipe


How many lines contain "river"?  We can count by piping into the word count tool `wc`.

In [20]:
!grep river siddhartha.txt | wc

    109    1365    7444


That's 109 matching lines, containing 1365 words and 7979 characters.  If you just wanted the lines by themselves, use `wc -l`:

In [21]:
!grep river siddhartha.txt | wc -l

109


What if we want to match both upper- and lower-case text?  Use `grep -i`:

In [22]:
!grep time siddhartha.txt | wc -l

166


In [23]:
!grep -i time siddhartha.txt | wc -l

167


How many lines in *Siddhartha* contain "other" (just lower-case)?  Start by using `grep` to extract lines that match the word "other" in `siddhartha.txt` and redirecting it to a file called `other-lines.txt`.

In [24]:
!grep other siddhartha.txt > other-lines.txt

In [25]:
%sc h_other = head -1 other-lines.txt
assert "other" in h_other

In [26]:
%sc t_other = tail -1 other-lines.txt
assert "other" in t_other

Now count up the lines in the file you created using wc.

In [27]:
!wc -l other-lines.txt

127 other-lines.txt


Your answer should be 127!

### Counting words with `grep`

By piping commands together we can do a lot of powerful things right at the command line.  Let's create a count of the most commonly occurring words in *Siddhartha*.  To do that, we could write a Python script that just counts words, but with the command line shell tools we only need to put a proper pipeline together and we can often accomplish tasks like this in one line.

First we need to split up the text lines into a word per line.  There are `grep` flags for that!

In [28]:
!cat siddhartha.txt | tr -sc '[:alpha:]' '[\n*]' | head -10


The
Project
Gutenberg
EBook
of
Siddhartha
by
Herman
Hesse
tr: write error: Broken pipe
tr: write error
cat: write error: Broken pipe


Now we need to sort them and count the unique tokens.  `sort` solves the first problem.

In [29]:
!cat siddhartha.txt | tr -sc '[:alpha:]' '[\n*]' | sort | head -10


a
a
a
a
a
a
a
a
a
sort: write failed: 'standard output': Broken pipe
sort: write error


And `uniq -c` solves the second problem.

In [30]:
!cat siddhartha.txt | tr -sc '[:alpha:]' '[\n*]' | sort | uniq -c | head -25

      1 
    953 a
     16 A
      2 abandon
      2 abandoned
      1 abide
      1 abilities
      3 ability
     41 able
      2 ablution
      3 ablutions
      1 abode
     83 about
      1 About
      6 above
      1 absorbed
      1 absorbing
      1 abstain
      4 abused
     11 accept
     13 accepted
      1 accepting
     10 access
      1 accessed
      1 accessible
uniq: write error: Broken pipe


But there's a catch... do you see it?

We need to convert all the words down into lower case so that we are correctly counting unique words.  There's another command, `tr`, for that.

In [32]:
!cat siddhartha.txt | tr -sc '[:alpha:]' '[\n*]' | tr '[:upper:]' '[:lower:]' | sort | uniq -c | head -25

      1 
    969 a
      2 abandon
      2 abandoned
      1 abide
      1 abilities
      3 ability
     41 able
      2 ablution
      3 ablutions
      1 abode
     84 about
      6 above
      1 absorbed
      1 absorbing
      1 abstain
      4 abused
     11 accept
     13 accepted
      1 accepting
     10 access
      1 accessed
      1 accessible
      2 accompanied
      2 accordance
uniq: write error: Broken pipe


...and if we want to know only the top 10 words in Siddhartha, we need to sort the output.

In [31]:
!cat siddhartha.txt | tr -sc '[:alpha:]' '[\n*]' | tr '[:upper:]' '[:lower:]' | sort | uniq -c | sort | head -10

      1 
    101 do
    104 t
    106 more
    108 any
    108 so
     10 access
     10 agree
     10 arms
     10 ask
sort: write failed: 'standard output': Broken pipe
sort: write error


But that sorts by character, not number.  Fortunately, `sort -n` does what we want.

In [33]:
!cat siddhartha.txt | tr -sc '[:alpha:]' '[\n*]' | tr '[:upper:]' '[:lower:]' | sort | uniq -c | sort -n | head -10

      1 
      1 abide
      1 abilities
      1 abode
      1 absorbed
      1 absorbing
      1 abstain
      1 accepting
      1 accessed
      1 accessible
sort: write failed: 'standard output': Broken pipe
sort: write error


But that's the wrong end of the list!  Two ways to fix that:  (a) use `tail` instead of `head`; (b) use `sort -rn`, which will sort in reverse order.  Let's try the latter.

In [34]:
!cat siddhartha.txt | tr -sc '[:alpha:]' '[\n*]' | tr '[:upper:]' '[:lower:]' | sort | uniq -c | sort -rn | head -10

   2221 the
   1434 and
   1225 to
   1107 of
    969 a
    960 he
    708 his
    686 in
    540 you
    524 had
sort: write failed: 'standard output': Broken pipe
sort: write error


Let's try another text.

Download Alice in Wonderland from https://s3.amazonaws.com/2017-dmfa/assignment-1/alice.txt

In [35]:
!wget https://s3.amazonaws.com/2017-dmfa/assignment-1/alice.txt

--2017-09-08 05:06:32--  https://s3.amazonaws.com/2017-dmfa/assignment-1/alice.txt
Resolving s3.amazonaws.com (s3.amazonaws.com)... 54.231.32.122
Connecting to s3.amazonaws.com (s3.amazonaws.com)|54.231.32.122|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 167515 (164K) [text/plain]
Saving to: ‘alice.txt’

alice.txt           100%[===================>] 163.59K  --.-KB/s    in 0.009s  

2017-09-08 05:06:32 (18.6 MB/s) - ‘alice.txt’ saved [167515/167515]



In [36]:
assert 'alice.txt' in os.listdir('.')

NameError: name 'os' is not defined

Take a look at the next cell.  Will it find the top 25 unique words in *Alice in Wonderland* successfully?

In [37]:
!cat alice.txt | tr -sc '[:alpha:]' '[\n*]' | sort | uniq -c | head -25

    672 a
     18 A
      2 abide
      1 able
    101 about
      1 About
      3 above
      1 absence
      2 absurd
      1 accept
      1 acceptance
      2 accepted
      1 accepting
     10 access
      1 accessed
      1 accessible
      2 accident
      1 accidentally
      2 accordance
      1 account
      1 accounting
      1 accounts
      1 accusation
      1 accustomed
      1 ache
uniq: write error: Broken pipe


Describe what needs to be done to the previous cell to get it to work correctly.  **Describe it using words**, explaining the issues, rather than using shell commands!

** The sort should be descending and by number instead of alphabetic

Okay, now implement your solution using shell commands with a pipeline.

In [41]:
! cat alice.txt | tr -sc '[:alpha:]' '[\n*]' | tr '[:upper:]' '[:lower:]' | sort | uniq -c | sort -rn |  head -25

   1818 the
    940 and
    809 to
    690 a
    631 of
    610 it
    553 she
    545 i
    481 you
    462 said
    431 in
    403 alice
    358 was
    330 that
    274 as
    248 her
    228 with
    227 at
    219 s
    218 t
    204 on
    200 all
    181 this
    179 for
    178 had
sort: write failed: 'standard output': Broken pipe
sort: write error
